src/err/mod.rs

```diff
@@ -0,0 +1,568 @@
+    /// });
+    /// ```
+    pub fn ptraceback<'py>(&'py self, py: Python<'py>) -> Option<&'py PyAny> {
+        self.normalized(py)
+            .ptraceback
+            .as_ref()
+            .map(|obj| obj.as_ref(py))
+    }
+
+    /// Gets whether an error is present in the Python interpreter's global state.
+    #[inline]
+    pub fn occurred(_: Python) -> bool {
+        unsafe { !ffi::PyErr_Occurred().is_null() }
+    }
+
+    /// Retrieves the current error from the Python interpreter's global state.
+    ///
+    /// The error is cleared from the Python interpreter.
+    /// If no error is set, returns a `SystemError`.
+    ///
+    /// If the error fetched is a `PanicException` (which would have originated from a panic in a
+    /// pyo3 callback) then this function will resume the panic.
+    pub fn fetch(py: Python) -> PyErr {
+        unsafe {
+            let mut ptype: *mut ffi::PyObject = std::ptr::null_mut();
+            let mut pvalue: *mut ffi::PyObject = std::ptr::null_mut();
+            let mut ptraceback: *mut ffi::PyObject = std::ptr::null_mut();
+            ffi::PyErr_Fetch(&mut ptype, &mut pvalue, &mut ptraceback);
+
+            let err = PyErr::new_from_ffi_tuple(py, ptype, pvalue, ptraceback);
+
+            if ptype == PanicException::type_object(py).as_ptr() {
+                let msg: String = PyAny::from_borrowed_ptr_or_opt(py, pvalue)
+                    .and_then(|obj| obj.extract().ok())
+                    .unwrap_or_else(|| String::from("Unwrapped panic from Python code"));
+
+                eprintln!(
+                    "--- PyO3 is resuming a panic after fetching a PanicException from Python. ---"
+                );
+                eprintln!("Python stack trace below:");
+                err.print(py);
+
+                std::panic::resume_unwind(Box::new(msg))
+            }
+
+            err
+        }
+    }
+
+    /// Creates a new exception type with the given name, which must be of the form
+    /// `<module>.<ExceptionName>`, as required by `PyErr_NewException`.
+    ///
+    /// `base` can be an existing exception type to subclass, or a tuple of classes
+    /// `dict` specifies an optional dictionary of class variables and methods
+    pub fn new_type<'p>(
+        _: Python<'p>,
+        name: &str,
+        base: Option<&PyType>,
+        dict: Option<PyObject>,
+    ) -> NonNull<ffi::PyTypeObject> {
+        let base: *mut ffi::PyObject = match base {
+            None => std::ptr::null_mut(),
+            Some(obj) => obj.as_ptr(),
+        };
+
+        let dict: *mut ffi::PyObject = match dict {
+            None => std::ptr::null_mut(),
+            Some(obj) => obj.as_ptr(),
+        };
+
+        unsafe {
+            let null_terminated_name =
+                CString::new(name).expect("Failed to initialize nul terminated exception name");
+
+            NonNull::new_unchecked(ffi::PyErr_NewException(
+                null_terminated_name.as_ptr() as *mut c_char,
```

Comment:
Does this take ownership of the `CString` or does `PyErr_NewException` copy the string? Else, I'm not sure how the pointer stays valid after `null_terminated_name` goes out of scope.

Reply:
It copies the string. (I believe this is the norm for all C APIs - it's pretty scary to steal ownership of things in C unless clearly documented, lest double-free!)

Reply:
Knowing this could also have some implications for CStrings in other interfaces like docstrings